src/asymmetric_crypto/rsa/private_key.rs

```diff
@@ -0,0 +1,151 @@
+use digest::{Digest, DynDigest};
+use pkcs8::SecretDocument;
+use rand_chacha::rand_core::CryptoRngCore;
+use rand_core::RngCore;
+use rsa::traits::PublicKeyParts;
+use zeroize::{ZeroizeOnDrop, Zeroizing};
+
+use crate::{
+    key_unwrap, pkcs8_fix, CryptoCoreError, PrivateKey, RsaKeyLength, RsaKeyWrappingAlgorithm,
+    RsaPublicKey,
+};
+
+#[derive(Hash, Clone, Debug, PartialEq, Eq)]
+pub struct RsaPrivateKey(rsa::RsaPrivateKey);
+
+impl RsaPrivateKey {
+    /// Generate a random private key
+    pub fn new<R: CryptoRngCore>(
+        rng: &mut R,
+        key_length: RsaKeyLength,
+    ) -> Result<Self, CryptoCoreError> {
+        Ok(Self(rsa::RsaPrivateKey::new(
+            rng,
+            (key_length as i32) as usize,
+        )?))
+    }
+
+    /// Get the key length which is the modulus size in bits
+    pub fn key_length(&self) -> RsaKeyLength {
+        match self.0.n().bits() {
+            2048 => RsaKeyLength::Modulus2048,
+            3072 => RsaKeyLength::Modulus3072,
+            4096 => RsaKeyLength::Modulus4096,
+            _ => panic!("Invalid RSA key length; this should never happen"),
+        }
+    }
+}
+
+/// Key wrapping support
+impl RsaPrivateKey {
+    /// Unwrap a key
+    pub fn unwrap_key(
+        &self,
+        wrapping_algorithm: RsaKeyWrappingAlgorithm,
+        encrypted_key_material: &[u8],
+    ) -> Result<Zeroizing<Vec<u8>>, CryptoCoreError> {
+        Ok(match wrapping_algorithm {
+            RsaKeyWrappingAlgorithm::Pkcs1v1_5 => Zeroizing::from(
+                self.0
+                    .decrypt(rsa::Pkcs1v15Encrypt, encrypted_key_material)?,
+            ),
+            RsaKeyWrappingAlgorithm::OaepSha256 => {
+                ckm_rsa_pkcs_oaep::<sha2::Sha256>(self, encrypted_key_material)?
+            }
+            RsaKeyWrappingAlgorithm::OaepSha1 => {
+                ckm_rsa_pkcs_oaep::<sha1::Sha1>(self, encrypted_key_material)?
+            }
+            RsaKeyWrappingAlgorithm::OaepSha3 => {
+                ckm_rsa_pkcs_oaep::<sha3::Sha3_256>(self, encrypted_key_material)?
+            }
+            RsaKeyWrappingAlgorithm::Aes256Sha256 => {
+                ckm_rsa_aes_key_unwrap::<sha2::Sha256>(self, encrypted_key_material)?
+            }
+            RsaKeyWrappingAlgorithm::Aes256Sha1 => {
+                ckm_rsa_aes_key_unwrap::<sha1::Sha1>(self, encrypted_key_material)?
+            }
+            RsaKeyWrappingAlgorithm::Aes256Sha3 => {
+                ckm_rsa_aes_key_unwrap::<sha3::Sha3_256>(self, encrypted_key_material)?
+            }
+        })
+    }
+}
+
+impl PrivateKey for RsaPrivateKey {
+    type PublicKey = RsaPublicKey;
+
+    fn public_key(&self) -> Self::PublicKey {
+        self.0.to_public_key().into()
```

Comment:
If you make the `RsaPublicKey`'s internal `pub(crate)`, you can avoid dynamic dispatch:
```suggestion
        RsaPublicKey(self.0.to_public_key())
```